src/test/java/com/google/sps/managers/DatastoreManagerTest.java

```diff
@@ -0,0 +1,301 @@
+// limitations under the License.
+
+package com.google.sps.managers;
+import com.google.sps.data.*;
+import com.google.sps.utils.*;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.gson.Gson;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import java.util.ArrayList;
+import com.google.gson.reflect.TypeToken;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class DatastoreManagerTest {
+  
+  DatastoreManager datastoreManager;
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  
+  @Before
+  public void setUp() {
+    helper.setUp();
+    datastoreManager = new DatastoreManager();
+    Gson gson = new Gson();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void createOrReplaceUserAndRetrieveTest() {
+    User expected;
+    User actual;
+
+    ArrayList<String> gamesCreated = new ArrayList<String>();
+    gamesCreated.add("game1");
+    gamesCreated.add("game2");
+    gamesCreated.add("game3");
+    gamesCreated.add("game4");
+    
+    //Makes use of the User class defined previously to build a single user for testing purposes
+    User.Builder user = new User.Builder("0001").setUsername("user1").setFirstName("Ade").setLastName("Ademiluyi");
+    user.setProfilePictureUrl("images/defaultProfilePicture.jpg").setGamesCreated(gamesCreated);
+    user.setGamesCompletedWithTime(new ArrayList<Pair<String, Long>>());
+    expected = user.build();
+
+    datastoreManager.createOrReplaceUser(expected);
+
+    actual = datastoreManager.retrieveUser("0001");
+
+    Assert.assertEquals(expected.getUserID(), actual.getUserID());
```

Comment:
Probably also check that the other fields are also preserved, because you already went through the effort of setting all of them.

Reply:
sure